Enable developers to run pod from terminal directly and skip pod install if possible.

kangwang1988 opened the pull request:
Currently.You have to run cocoapods from the flutter side, in which it will pass environment variables containing the location of the flutter framework. However, we want a standard workflow, i.e., you can run pod install from flutter, you can also run it from the terminal.

Commit:
update_script

kangwang1988 committed:
[Changes]:Enable run pod from terminal directly.

Commit:
[Changes]:Enable run pod from terminal directly.

kangwang1988 committed:
[Changes:]Change test for cocoapods to make it work properly after modification for the workflow.

Commit:
[Changes:]Change test for cocoapods to make it work properly after modification for the workflow.

cbracken committed:
Apply media bottom padding to CupertinoTabBar (#13372)

This adjusts the CupertinoTabBar height in the presence of bottom
padding. On the iPhone X, this will increase the height to account for
the home indicator widget.

bkonyi committed:
Flutter engine roll 1/2 for 2017/12/08

bkonyi committed:
Flutter engine roll 2/2 (with Dart SDK roll) for 2017/12/08

kangwang1988 committed:
[Changes]:Check if you need to run pod install.
In the cases below, the pod install will not be skipped.
1.The plugin has changed (add/update/delete)
2.The flutter.framework has changed (debug/release/profile)
3.The podfile.lock doesn't exists
4.The Pods/manifest.lock doesn't exists
5.The podfile.lock doesn't match Pods/manifest.lock.

Commit:
[Changes]:Check if you need to run pod install.
In the cases below, the pod install will not be skipped.
1.The plugin has changed (add/update/delete)
2.The flutter.framework has changed (debug/release/profile)
3.The podfile.lock doesn't exists
4.The Pods/manifest.lock doesn't exists
5.The podfile.lock doesn't match Pods/manifest.lock.

Commit:
[Chages]:
Fix a analyze problem.

kangwang1988 committed:
[Chages]:
Fix a analyze problem.

devoncarew committed:
some wording tweaks to the doctor text (#13356)

kevmoo committed:
flutter_tools: ensure the context value passed to pub is consistent (#13366)

Adds a class `PubContext` with a fixed set of allowed values
Make it clear these values should not be changed casually

Fixed one inconsistency already: update_packages vs update_pkgs
Provide more information for verify calls
Eliminate `ctx_` prefix.

cbracken committed:
Roll engine to 033daf152d78919101c0022f98fb18fd830284c0 (#13387)

This picks up a single commit: flutter/engine#4418, which fixes #12585.

Hixie committed:
Update gallery for https://github.com/flutter/flutter/pull/13195 (#13361)

gspencergoog committed:
Fixing the pub get test so that it doesn't care what the flutter root looks like. (#13385)

goderbauer committed:
Do not show handles when changing text selection via keyboard (#13338)

This contains a breaking API change to `SelectionChangedHandler`.

This is part of making text input accessible on Andriod (https://github.com/flutter/flutter/issues/12785).

xster committed:
Add helper function to create cylindrical projection matrices (#13376)

* Add a cylindrical projection helper matrix

* specify tangential

* more doc clarifications

* reshuffle arguments

* more code comments

* add some sanity tests

* review

* review 2

* added one more caveat for consideration in docs

cbracken committed:
Roll engine to f888186e50f20fdb49ceea0dae74b6443a22ddaa (#13402)

Rolls:
flutter/engine#4411
flutter/engine#4416
flutter/engine#4420
flutter/engine#4421
flutter/engine#4423

yjbanov committed:
Emulate text entry in FlutterDriver (#13373)

* Emulate text entry in FlutterDriver

* document enterText behavior

* remove the unnecessary composint TextRange

yjbanov committed:
link from CONTRIBUTING.md to various build infra docs (#13399)

yjbanov committed:
support disabling text entry emulation (#13410)

szakarias committed:
Increase connection timeout (#13413)

szakarias committed:
Increase connection timeout further (#13415)

szakarias committed:
Revert "Increase connection timeout further (#13415)" (#13417)

This reverts commit f40a9398a5547c653774570bc442104789d12c34.

jakobr-google committed:
Make artifacts URLs configurable. (#13380)

Add support for configuring the base storage URL for Flutter's
artifacts. If FLUTTER_STORAGE_BASE_URL is set, use it instead of
storage.googleapis.com.

The pub server can be overridden by setting PUB_HOSTED_URL.

szakarias committed:
Revert "Increase connection timeout" (#13418)

devoncarew committed:
route device issue diagnostics to flutter doctor (#13346)

* route device issue diagnostics to flutter doctor

* review comments

* review comments

jacob314 committed:
Change enum properties to use camel case instead of dash separated names

Make regular Alignment values look more like enumProperties while
leaving more complex AlignmentDirectional values unchanged.

yjbanov committed:
more setTextEntryEmulation docs; isRegistered check (#13420)

cbracken committed:
Use greater of viewInsets, padding for Scaffold bottom padding (#13423)

Scaffold bottom padding now applies the maximum of window
viewInsets.bottom (typically used for iOS safe areas) and padding.bottom
(typically used for keyboard height).

kevmoo committed:
(cleanup) standardize on shouldRunPub and usesPubOptions() in commands (#13427)

kevmoo committed:
Revert "(cleanup) standardize on shouldRunPub and usesPubOptions() in commands (#13427)" (#13430)

This reverts commit 9315a79b11dfcd82ffd1af9da7451b2e29e71253.

There is an unfortunate interaction with verifyThenRunCommand

Hixie committed:
Move image logic from services/ to painting/. (#13409)

This allows the scheduler library to depend on the services library
and the painting library to depend on the scheduler library without
the services library having to depend on the scheduler library.

While I was at it I also cleaned up some of the binding logic: the
licenses logic can now be overridden (and the test library does so),
and the image cache can now be overridden as well.

cbracken committed:
Update scaffold test bottom bar height (#13429)

This patch contains no framework changes.

This change adjusts a test bottom navigation bar height to be greater
than the media bottom padding.

This change is pre-factoring to clarify diffs in an upcoming change that
will expose bottom padding to the bottom navigation bar in order to
allow it to visually adapt its appearance in the presence of bottom
padding such as thatexposed for the iPhone X home screen indicator.
Since the bottom padding was previously greater than the height of the
bottom navigation bar, the height of insideBottomNavigation bar was
driven to 0 when padding is exposed.

jason-simmons committed:
Fix calculation of downstream caret location at the end of a string (#13426)

cbracken committed:
Extract a shared MediaQueryData for padding and viewInsets (#13434)

Also adds a TODO that should have landed with flutter/flutter#13423.

cbracken committed:
Scaffold resizeToAvoidBottomPadding uses view insets (#13437)

Now that keyboard height is modelled as a (bottom) view inset, migrate
scaffold bottom resizing to use view insets instead of bottom padding,
which, after an engine roll, will only be used for safe areas.

Until the aforementioned engine roll, the keyboard height is still
included in both bottom padding and view insets. As such
resizeToAvoidBottomPadding still drives bottom padding removal until
that roll lands.

Renames _ScaffoldLayout.bottomPadding to bottomViewInset

xster committed:
Remove dangling MediaQuery paddings for non-fullscreen dialogs - 2 (#13438)

* Remove MediaQuery padding for various PopupRoutes that don't extend to the edge of the screen

* bottom sheet doesn't actually need handling. Scaffold does it already. Add test for others.

cbracken committed:
Apply media bottom padding in BottomNavigationBar (#13431)

Updates scaffold to expose bottom padding to the associated
BottonNavigationBar (if present). BottomNavigationBar is now responsible
for adjusting itself to account for bottom padding.

This change is necessary to support the updated BottomNavigationBar
layout for the iPhone X.

konifar committed:
Remove unnecessary error message when cocopods is not available (#13241)

* Remove unnecessary error message

* Add AUTHOR

goderbauer committed:
a11y cursor movement (#13405)

cbracken committed:
Roll engine to 1cdaab88e33f564eee8455c75479706293f2b655 (#13441)

Includes:
flutter/engine#4422 (Dart SDK roll)

cbracken committed:
Roll engine to 8e39760168e7669a023d1746823d9ab6db4e5df8 (#13442)

Includes:
flutter/engine#4427
flutter/engine#4428

jcollins-g committed:
Upgrade dartdoc to 0.15.0+1 (#13432)

* Update dartdoc to 0.15.0+1 for Flutter.

* Minimize change

eseidelGoogle commented:
@xster can you review this and/or help bring this over the line for customer:gold.

xster committed:
Let translucent Cupertino bars have its scaffold children automatically pad their heights - second try (#13440)

* Let translucent Cupertino bars have its scaffold children automatically pad their heights (#13194)

* Let lists automatically add sliver padding from media query. Translucent nav and tab bars leave behind media query paddings in scaffolds.

* tests

* const lint

* Rename base abstract class to generalized ObstructingPreferredSizeWidget

* review

* More docs and comments from #13317

zanderso committed:
[flutter_tools][plugins] generated registerWith guards against re-registration (#13384)

bkonyi committed:
Enabling 'flutter test' for Windows (woohoo!).

yjbanov committed:
Make time picker accessible (#13152)

* make time picker accessible

* use new CustomPaint a11y API

* flutter_localizations tests; use bigger distance delta

* fix am/pm control; selected values

* fix translations; remove @mustCallSuper in describeSemanticsConfiguration

* exclude AM/PM announcement from iOS as on iOS the label is read back automatically

Hixie committed:
Fix tapping on a test in flutter run (#13397)

Also:

* Remove find.byIcon since it's identical to find.icon. (I sent mail to flutter-dev about this.)

* Fix IconData's operator== and hashCode, which had not been updated when we added fields.

* Add the byTooltip finder to the list of suggested finders.

* Make the suggested Key finder prettier.

bkonyi committed:
Removed special cases for '.exe'.

bkonyi committed:
Removed dead code.

bkonyi committed:
Removed additional dead code.

gspencergoog committed:
Make create command use pub in offline mode. (#13462)

In order to allow offline usage of the create command we need to be able to tell pub not to try and contact the server to see if new versions are available. Since all the versions that create could want are pinned, it shouldn't need to check for new versions.

sethladd committed:
docs to redirect a user to dart:io.HttpClientOverrides (#13465)

athomas committed:
Make it possible to specify multiple test arguments in test.dart (#13383)

This removes the FLUTTER_TEST_ARGS environment variable handling.

Hixie committed:
Have the framework in charge of scheduling frames. (#13344)

...instead of the engine.

xster commented:
Thanks for your contribution. I started reviewing this and then thought about what could be the root cause since repeated runs on `pod install` should have more or less no cost. And it seems like your main concern is repeated builds on the same machine. 

We may be able to address the underlaying issue in #13467. Based on https://github.com/CocoaPods/CocoaPods/issues/4128, it's possible it's slow for you guys in China only because of stats collection in CocoaPods.

kangwang1988 commented:
@xster 
In our team, the problems brought by cocoapods(pod install) are as follows:
1.It takes time to complete a full repo-update. Our repos not only contain the official github one, but also  our private one. Not to mention the slow/unstable network connection to github.
2.As cocoapods re-generate the whole Pods.project each time for good, when our project contains many dependencies, the generating pod project process could cost at least 1 min every time.

This PR mainly focuses in two sides:
1.Enable to run pod install directly.
Developers are used to run their pod install from terminal. However current implementation in flutter_tools makes it hard. Also, when debugging a cocoapods problem or do some kind of customization cocoapod process, it will simplify the process and save efforts when following the standard way.
2.Skip the pod install if possible
There are two reasons.
a.As mentioned above, the cocoapods process is time costly.In our intranet(with good network condition and good access to github), even after a lot optimization, the process costs 1+ mins each time when the dependency is complicated(especially in a company where there are plenty of middlewares).
b.We specify our dependency for certain commit/tag for safety and only run the pod install process when we clearly know we need to update some of them.

cbracken committed:
Remove bottom padding in SliverAppBar (#13471)

Removes bottom padding from the child AppBar in SliverAppBar. Scaffold
already does this for its own app bars, but sliver app bars in the body
should also apply the bottom padding removal.

Fixes flutter/flutter#13458

xster commented on packages/flutter_tools/lib/src/ios/mac.dart:
```diff
@@ -251,15 +251,24 @@ Future<XcodeBuildResult> buildXcodeProject({
-      isSwift: app.isSwift,
-    );
-
+        appIosDir: appDirectory,
```
Nit, 2 spaces for parameters

xster commented on packages/flutter_tools/lib/src/ios/mac.dart:
```diff
@@ -352,7 +361,31 @@ Future<XcodeBuildResult> buildXcodeProject({
 }
 
-Future<Null> diagnoseXcodeBuildFailure(XcodeBuildResult result, BuildableIOSApp app) async {
+//Write the flutter_framework path to appPath/ios/.flutter_framework, which will be read by the podfile later.
```
If the above works for you, you won't need to create the next 2 methods

xster commented on packages/flutter_tools/lib/src/ios/cocoapods.dart:
```diff
@@ -139,6 +138,28 @@ class CocoaPods {
     }
   }
 
+  //Check if you need to run pod install.
```
1 space after each //

xster commented on packages/flutter_tools/lib/src/ios/cocoapods.dart:
```diff
@@ -139,6 +138,28 @@ class CocoaPods {
   }
 
+  //Check if you need to run pod install.
+  //In the cases below, the pod install will not be skipped.
```
Avoid double negations. Just say pod install will run if any of the below is true

xster commented on packages/flutter_tools/lib/src/ios/cocoapods.dart:
```diff
@@ -139,6 +138,28 @@ class CocoaPods {
 
+  //Check if you need to run pod install.
+  //In the cases below, the pod install will not be skipped.
+  //1.The plugin has changed (add/update/delete)
```
Plugin is plural. e.g. 'Any plugins changed'

xster commented on packages/flutter_tools/lib/src/ios/cocoapods.dart:
```diff
@@ -139,6 +138,28 @@ class CocoaPods {
+  //3.The podfile.lock doesn't exists
+  //4.The Pods/manifest.lock doesn't exists
+  //5.The podfile.lock doesn't match Pods/manifest.lock.
+  bool _checkIfSkipPodInstall(String appDir, bool pluginOrFlutterPodChanged) {
```
Avoid double negation in method name too (i.e. checking ! skip).

xster commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -77,6 +77,17 @@ List<Plugin> _findPlugins(String directory) {
   return plugins;
 }
 
+/// Read the content of .fluter-plugins and return as string.
```
if the above works for you, you may not need this either

xster commented on packages/flutter_tools/templates/cocoapods/Podfile-objc:
```diff
@@ -1,30 +1,45 @@
+  # Pods for Runner
   # Flutter Pods
-  pod 'Flutter', :path => ENV['FLUTTER_FRAMEWORK_DIR']
+  flutter_pods_ary = parse_KV_file(".flutter-framework")
```
Reading directly out of Generated.xcconfig would be nice without creating a separate file.

Semantically, it fits the natural hierarchy too by create a single source of hierarchical truth. CocoaPods shouldn't work anyway if Generated.xcconfig doesn't exist (meaning the user never ran flutter build/run before or the build is outdated).

xster reviewed (commented):
Thanks for the patch. The general approach looks right to me. Please add a test to confirm the new behaviour

Some questions on your last comment:

1- My understanding is that CocoaPods stopped invoking pod repo update automatically during pod install since 2.0. Flutter tools explicitly never invokes pod setup or pod repo update. Is this an orthogonal problem.
2- I'm slightly surprised. I was under the impression that once CocoaPods generates the new Podfile.lock from the new Podfile and if it matches manifest.lock, it'll just not do anything and not touch anything under the existing Pods. But I might be wrong. Do the Podfile.lock change for you guys between flutter runs?

b. Ha, that's a good point. I think you hit on one of our issues with #9502. Would that problem go away for you guys if you can version control Podfile.lock?
c. That's definitely an issue. I guess it's triggered to point 2 above which I'm surprised by

cbracken committed:
Roll engine to edb0201fa258d6a2c05899c185a4e5cba0dc26a3 (#13473)

Includes:
flutter/engine#4429
flutter/engine#4415
flutter/engine#4431
flutter/engine#4432
flutter/engine#4401
flutter/engine#4434
flutter/engine#4435
flutter/engine#4436

mjohnsullivan committed:
Fix template for the driver test file created by 'flutter create' (#13472)

mehmetf committed:
Fix for incorrect transparent -> DividerColor animation for ExpansionTile (#13449)

* Fix for incorrect transparent -> DividerColor animation for ExpansionTile

* Add comments and tests. Fix background color too.

* Fix analyzer warnings

kangwang1988 committed:
[Modified]:Refactor codes & documentation according to the flutter team suggestions.

Commit:
[Modified]:Refactor codes & documentation according to the flutter team suggestions.

kangwang1988 committed:
[Modified]:Add test for skip pod install if possible.

Commit:
[Modified]:Add test for skip pod install if possible.

kangwang1988 commented:
@xster 
Currently, I'm using 1.3.1.
1.In 1.3.1, the repo-update part you said is right. 
`Note: as of CocoaPods 1.0, `pod repo update` does not happen on `pod install` by default.`  referring to https://github.com/CocoaPods/CocoaPods/issues/5260.
The prior idea I hold that pod install will trigger repo update is a designed feature. It comes from our company-customized cocoapod plugin. It works in a non-standard way where you can only skip the pod repo update process when a --no-repo-update option is given.
2.I dig more and found that, even the Podfile.lock matches Pods/Manifest.lock, cocoapods will also modify Pods/Manifest.lock and Pods/Pods.xcodeproj/project.pbxproj. In my opinion,this is mainly  because of the consideration of the existence of `Development Pods`.There was a bad experience when I seperated one project into many development pods while they have to be changed frequently. Add/Delete a source or resource will require the pod install process to get right project file.As file references,libraries,target_dependencies,etc(Refer to gems/cocoapods-1.3.1/lib/cocoapods/installer/xcode/pods_project_generator.rb for more) have been modified in this process, the whole project will be recompiled.

kangwang1988 committed:
Merge branch 'master' into master

kangwang1988 committed:
[Modified]:Fix analyze issues.

Commit:
[Modified]:Fix analyze issues.

Commit:
Merge branch 'master' of github.com:kangwang1988/flutter

* 'master' of github.com:kangwang1988/flutter: (49 commits)
  Fix for incorrect transparent -> DividerColor animation for ExpansionTile (#13449)
  Fix template for the driver test file created by 'flutter create' (#13472)
  Roll engine to edb0201fa258d6a2c05899c185a4e5cba0dc26a3 (#13473)
  Remove bottom padding in SliverAppBar (#13471)
  Have the framework in charge of scheduling frames. (#13344)
  Make it possible to specify multiple test arguments in test.dart (#13383)
  docs to redirect a user to dart:io.HttpClientOverrides (#13465)
  Make create command use pub in offline mode. (#13462)
  Fix tapping on a test in flutter run (#13397)
  Make time picker accessible (#13152)
  [flutter_tools][plugins] generated registerWith guards against re-registration (#13384)
  Let translucent Cupertino bars have its scaffold children automatically pad their heights - second try (#13440)
  Upgrade dartdoc to 0.15.0+1 (#13432)
  Roll engine to 8e39760168e7669a023d1746823d9ab6db4e5df8 (#13442)
  Roll engine to 1cdaab88e33f564eee8455c75479706293f2b655 (#13441)
  a11y cursor movement (#13405)
  Remove unnecessary error message when cocopods is not available (#13241)
  Apply media bottom padding in BottomNavigationBar (#13431)
  Remove dangling MediaQuery paddings for non-fullscreen dialogs - 2 (#13438)
  Scaffold resizeToAvoidBottomPadding uses view insets (#13437)
  ...

jason-simmons committed:
Fix one of the text painter RTL tests (#13470)

xster commented on packages/flutter_tools/lib/src/ios/cocoapods.dart:
```diff
@@ -137,6 +138,28 @@ class CocoaPods {
   }
 
+  // Check if you need to run pod install.
+  // The pod install will run if any of below if true.
```
*if any of the below **is** true

xster commented on packages/flutter_tools/lib/src/ios/mac.dart:
```diff
@@ -268,6 +263,16 @@ Future<XcodeBuildResult> buildXcodeProject({
   );
 
+  if(hasFlutterPlugins){
+    final String curGeneratedXCConfig = readGeneratedXCConfig(app.appDirectory);
```
Don't abbreviate, just call it 'currentGeneratedXCConfig' or 'newGeneratedXCConfig'

xster commented on packages/flutter_tools/lib/src/ios/mac.dart:
```diff
@@ -268,6 +263,16 @@ Future<XcodeBuildResult> buildXcodeProject({
+  if(hasFlutterPlugins){
+    final String curGeneratedXCConfig = readGeneratedXCConfig(app.appDirectory);
+    await cocoaPods.processPods(
+        appIosDir: appDirectory,
```
nit: 2 spaces for paramaters

xster commented on packages/flutter_tools/lib/src/ios/mac.dart:
```diff
@@ -268,6 +263,16 @@ Future<XcodeBuildResult> buildXcodeProject({
+        appIosDir: appDirectory,
+        isSwift: app.isSwift,
+        pluginOrFlutterPodChanged:
+        (injectionResult.hasChanged ||
```
nit: 4 more spaces here for line continuations

xster commented on packages/flutter_tools/lib/src/ios/mac.dart:
```diff
@@ -352,7 +357,17 @@ Future<XcodeBuildResult> buildXcodeProject({
-Future<Null> diagnoseXcodeBuildFailure(XcodeBuildResult result, BuildableIOSApp app) async {
+String readGeneratedXCConfig(String appPath) {
+  String generatedXCConfig;
+  final String flutterPodWrapperPath =
```
Why is this called flutterPodWrapperPath? There's a bunch of general Flutter flags. Just call it generateXCConfigPath

xster commented on packages/flutter_tools/lib/src/ios/mac.dart:
```diff
@@ -352,7 +357,17 @@ Future<XcodeBuildResult> buildXcodeProject({
+  return generatedXCConfig;
+}
+
+Future<Null> diagnoseXcodeBuildFailure(
```
nit: I think you can just keep this in one line

xster commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -77,8 +77,10 @@ List<Plugin> _findPlugins(String directory) {
+// Return true if .flutter-plugins has changed, otherwise return false.
+bool _writeFlutterPluginsList(String directory, List<Plugin> plugins) {
   final File pluginsProperties = fs.file(fs.path.join(directory, '.flutter-plugins'));
+  final String priorFlutterPlugins = (pluginsProperties.existsSync()?pluginsProperties.readAsStringSync():null);
```
nit: spaces around ? and :

xster commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -88,6 +90,8 @@ void _writeFlutterPluginsList(String directory, List<Plugin> plugins) {
       pluginsProperties.deleteSync();
     }
   }
+  final String curFlutterPlugins = (pluginsProperties.existsSync()?pluginsProperties.readAsStringSync():null);
```
nit: don't abbreviate

xster commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -206,17 +210,25 @@ void _writeIOSPluginRegistry(String directory, List<Plugin> plugins) {
+  bool anyPlugin;
+  // True if plugins have changed.
+  bool hasChanged;
+  InjectPluginsResult({this.anyPlugin = false,this.hasChanged=false});
```
nit: let the constructor be on top, use : for defaults, like `this.anyPlugin: false`. Space after `,`

xster commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -206,17 +210,25 @@ void _writeIOSPluginRegistry(String directory, List<Plugin> plugins) {
 
+class InjectPluginsResult{
+  // True if any flutter plugin exists, otherwise false.
+  bool anyPlugin;
```
`hasPlugin`, also use `final` for immutable data

xster commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -206,17 +210,25 @@ void _writeIOSPluginRegistry(String directory, List<Plugin> plugins) {
 }
 
+class InjectPluginsResult{
+  // True if any flutter plugin exists, otherwise false.
```
nit: use `///`, it becomes documentation instead of comments

xster commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -206,17 +210,25 @@ void _writeIOSPluginRegistry(String directory, List<Plugin> plugins) {
+class InjectPluginsResult{
+  // True if any flutter plugin exists, otherwise false.
+  bool anyPlugin;
+  // True if plugins have changed.
```
True if plugins have changed since last build.

xster commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -206,17 +210,25 @@ void _writeIOSPluginRegistry(String directory, List<Plugin> plugins) {
     _writeIOSPluginRegistry(directory, plugins);
-  return plugins.isNotEmpty;
-}
+  return new InjectPluginsResult(anyPlugin: plugins.isNotEmpty,hasChanged:hasPluginsChanged);
```
nit: space after `,`, space after `:`

xster commented on packages/flutter_tools/templates/cocoapods/Podfile-swift:
```diff
@@ -1,32 +1,41 @@
-end
-
-target 'Runner' do
-  use_frameworks!
```
We still need this line

xster commented on packages/flutter_tools/templates/cocoapods/Podfile-swift:
```diff
@@ -1,32 +1,41 @@
+def parse_KV_file(file,seperator='=')
+  file_abs_path = File.expand_path(file)
+  if !File.exists? file_abs_path
+    puts file_abs_path+"Not exists"
```
We should add a message here saying `Generated.xcconfig` or `.flutter-plugins` expected but not present. If you're running `pod install` manually, make sure `flutter build` or `flutter run` is executed once first. 

Then abort instead of letting it potentially finish and have subtle bugs

xster commented on packages/flutter_tools/templates/cocoapods/Podfile-swift:
```diff
@@ -1,32 +1,41 @@
+
+target 'Runner' do
+  # Flutter Pods
+  flutter_pods_ary = parse_KV_file("./Flutter/Generated.xcconfig")
```
maybe generated_xcode_build_settings

xster commented on packages/flutter_tools/templates/cocoapods/Podfile-swift:
```diff
@@ -1,32 +1,41 @@
+  }
+
+  # Plugin Pods
+  plugin_pods_ary = parse_KV_file("../.flutter-plugins")
```
nit: Don't abbreviate. Just plugin_pods

xster reviewed (approved):
This is awesome! Thanks for making the changes. 

Also, please remember to sign the CLA in the link posted by the bot above and then reply `I signed it!` here in the comment section and I'll merge it for you.

fredriks committed:
Expose textAlign on TextFormField (#13414)

* Expose textAlign on TextFormField

Fixes #11404

* Added name to AUTHORS

* Added a test for TextFormWidget's textAlign

zanderso committed:
Roll engine to edaecdc8b8098f20512c23745b9d0b135a8ee56d (#13488)

Hixie committed:
Remove hack for support of old Hebrew language code. (#13309)

See https://github.com/flutter/flutter/issues/11907 and https://github.com/flutter/engine/pull/4411

kangwang1988 committed:
Refactor for style & documentation.

Commit:
Refactor for style & documentation.

kangwang1988 commented:
@xster  
Thank you for your comments on the coding style and documentation. I've a question. Some logic or parameter/method naming may be hard to analyze. However, as to  the format issue(like spaces after ? : in a ternary ), is there any tool which can help us locate and fix it? I tried flutter analyze but it seems not enough.

gspencergoog committed:
Make the create command call pub get online instead of offline. (#13505)

We thought it would be OK to call the create command offline-only, but it turns out that isn't a great idea because not all of the create template dependencies are cached by running flutter_tool.

This PR makes it go online by default when running pub get after creating a project, and adds an "--offline" flag to the create command so that offline users can use it offline if their cache is warm.

xster commented:
Unfortunately we're still in the process of potentially tweaking dartfmt a bit to let the process be automated in the future. For the time being, it's manual applications of the clauses listed at https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo.

gspencergoog committed:
Rolling alpha to 0.0.20 (#13511)

Hixie committed:
Tests for TextPainter.minIntrinsicWidth (#13515)

xster committed:
Remove top mediaquery padding from modal bottom sheets (#13497)

cbracken committed:
Roll engine to 6c6b709b2c4008e243dda3c148521c35e6e8b3df (#13522)

Includes:
flutter/engine#4442
flutter/engine#4445
flutter/engine#4444
flutter/engine#4448 (reverts flutter/engine#4444)
flutter/engine#4447

gspencergoog committed:
Updating SDK and package versions (#13524)

Updating SDK and package versions now that alpha has been rolled to 0.0.20

cbracken committed:
Roll engine to 4ea7075c1def9dce38f16a3c9bf0ed6fb6b2d441 (#13523)

Includes:
flutter/engine#4425

mehmetf committed:
Add test for interim state of the animation to make sure we are transitioning the border color the right way. (#13499)

Hixie committed:
Fix TableBorder bugs (#13495)

Add more tests, fix docs, fix painting bugs...

Fixes https://github.com/flutter/flutter/issues/12902

mehmetf committed:
Increase coverage collection timeout to 2 minutes. Leafy is hitting the limits of the previous timeout of 30 seconds for tests that touch a lot of code in the app. (#13527)

goderbauer committed:
Roll engine to 2bdb21985c5eb9250a7b744b9b4d606c5ba30945 (#13526)

goderbauer committed:
Add arguments to SemanticsActions; implement extend selection for a11y (#13490)

**This PR contains a breaking API change:**
`SemanticsConfiguration.addAction()` has been removed and replaces by action-specific setters (`onTap`, `onLongPress`, etc.) to take care of the fact that some action handlers (those, who take arguments) have different signatures.

mravn-google committed:
Support Android Gradle plugin v3.0.1 (#13492)

mravn-google committed:
Fix gradle plugin test (#13535)

yjbanov committed:
update APK path in devicelab tests (#13540)

* update APK path in devicelab tests

* add missing transitive dependency on archive

* use path used by flutter tools

szakarias committed:
Include a directory with Flutter assets (#12944)

szakarias committed:
Upgrade project.pbxproj to include flutter_assets (#13011)

szakarias committed:
Upgrade complex_layout project.pbxproj to include flutter_assets (#13544)

yjbanov committed:
mark flutter_gallery_ios__start_up as flaky (#13542)

yjbanov committed:
Revert engine (#13547)

* Revert "Include a directory with Flutter assets (#12944)"

This reverts commit 3af6b9cbf5787b4db6b4bad7494fe41addf5703d.

* Revert "Upgrade project.pbxproj to include flutter_assets (#13011)"

This reverts commit 08128cb29b28db00c2301f546bc59de33017074a.

* Revert "Upgrade complex_layout project.pbxproj to include flutter_assets (#13544)"

This reverts commit 35f1a04195e69afb5cc1e6588abad5476db3a5d8.

* mark complex_layout_ios__start_up as flaky

Hixie committed:
Update AnimatedDefaultTextStyle to match DefaultTextStyle (#13531)

We've added a bunch of non-animatable properties to DefaultTextStyle.
This adds them to AnimatedDefaultTextStyle so that you can use them
even when your text style is animating.

xster committed:
Add more doc to MediaQuery padding (#13525)

* add mediaquery docs

* review

jason-simmons committed:
Fixes to a text_painter_rtl whitespace test (#13551)

* Query selection rectangles for all three characters in the text
* Change the last character to be non-whitespace.  Minikin's line breaker will
  truncate whitespace characters at the end of a line.

goderbauer committed:
Verify that committed GeneratedPluginRegistrant are up-to-date (#13541)

cbracken committed:
Apply media padding in Cupertino nav bar (#13550)

Applies left and right media padding to Cupertino navigation bar large
titles.

cbracken committed:
Apply media padding in ListTile (#13545)

Applies media padding (e.g. iOS safe area insets) around the list tile
children inside the InkWell.

Also adds safe area around gallery section titles for consistency with
surrounding list tiles.

xster committed:
Add more checks to scheduleWarmUpFrame (#13552)

* Add more checks to scheduleWarmUpFrame

* review

* add test

xster committed:
Add some docs and warnings to PageController.page (#13509)

* partial

* add doc and warning

* add correction

* review

* fix test

mravn-google committed:
Simplify project template (#13558)

pq committed:
Updated flutter_tools test running instructions. (#13533)

mravn-google committed:
Refresh template instances (#13563)

mravn-google committed:
Revert "Refresh template instances (#13563)" (#13565)

This reverts commit 05d6b08ffe12bed1328f65d911440ddab55751a2.

szakarias committed:
Write flutter_assets directory on Windows (#13564)

szakarias committed:
Reapply flutter_assets (#13567)

* Revert "Revert engine (#13547)"

This reverts commit ee65db11d4618620ddc286f4618b707239205136.

* Keep flaky mark

pq committed:
Update plugin test template. (#13516)

* Update plugin test template.

* Add flutter test run verification.

* Tweak and test runs for package projects.

amirh committed:
Pin the test package to 0.12.26 (#13546)

amirh committed:
flutter update-packages (#13571)

* flutter update-packages

* update plugin registrant for flutter_gallery

crelier committed:
Roll engine to e0d19e2897b6c4c8c48f05d23a2b90bd32ab5d49 (#13570)

amirh committed:
when plugins are out of date, show the command line to update (#13572)

xster committed:
Document various 'alignment' members so it's clear how to concretely specify it. (#13577)

goderbauer committed:
Add a11y label to modal barrier (#13543)

Hixie committed:
Allow the ErrorWidget to be overridden. (#13578)

Fixes https://github.com/flutter/flutter/issues/10695

cbracken committed:
Apply T, L, R media padding in Cupertino nav bar (#13580)

Applies top, left, right media padding for iOS safe area insets. This is
a follow-up to the media padding applied to large titles in
flutter/flutter#13550.

cbracken committed:
Center Shrine demo title with centerTitle (#13586)

Rather than use a Center widget, center the title using AppBar's
centerTitle property. This ensures the title is horizontally centred
with respect to the screen rather than centred in the space between the
leading and trailing app bar widgets, which are asymmetrical in Shrine.

cbracken committed:
Apply media padding in Cupertino Navigation demo (#13585)

Adds safe areas around:
1. The body of the colour swatch view (tab 1)
2. The chat header (tab 2)
3. The chat bubbles (tab 2)
4. The 'Sign In' launcher button (tab 3)

cbracken committed:
Apply media padding in Color, Typography demos (#13592)

yjbanov committed:
make date picker accessible (#13502)

* make date picker accessible

* make test file lookup location-independent

* address some comments

* always wrap in IgnorePointer

* no bitmasks for flags and actions

* recommend List<*>

mkustermann committed:
Roll engine to d945976b04cc82d32c9691dae1fc8e4955222b3d (#13605)

cbracken committed:
Apply media padding to Gallery Material Design demos (#13593)

Applies media padding (e.g. iPhone X safe area insets) to the Material Design demos in the Gallery.

Covers the following demos:
* Buttons (via the TabbedComponentDemoScaffold change)
* Cards
* Expansion panels
* Grid list
* Icons
* Images (via the TabbedComponentDemoScaffold change)
* Page Selector
* Progress Indicator
* Scrollable tabs
* Selection controls (via the TabbedComponentDemoScaffold change)
* Snack bar
* Tabs
* Text fields
* Tooltips

Fixes #13594

cbracken committed:
Require Xcode 9.0 or later for iOS builds (#13608)

We now require Xcode 9.0, which ships with the iOS 11 SDK, for iOS
builds. This change does not affect the minimum supported iOS deployment
target, which remains iOS 8.

The iOS 11 SDK adds support for safe area insets, which model the status
bar, iPhone X home indicator, and symmetric horizontal insets to avoid
the iPhone X sensor housing when in landscape mode.

cbracken committed:
Revert "Require Xcode 9.0 or later for iOS builds (#13608)" (#13612)

Our devicelab machines have not yet been upgraded to Xcode 9. Once this
has been done, I'll re-land this.

This reverts commit 0fac5f01a2469b2658a64da139884b337de5f9dd.

amirh committed:
keep recursing past directories with pubspec.yaml when gathering packages (#13573)

cbracken committed:
Apply media padding in complex layout benchmark drawer header (#13610)

By default BoxScrollView (and hence ListView, which is a subclass)
padding is the media padding along its scroll axis in order to avoid
placing list items within areas where user interaction should be
minimised -- e.g. under the status bar, or in and around the iPhone X
notch in landscape mode.

In cases where a list item should occupy the padding area, developers
should set the ListView padding to EdgeInsets.zero so as not to pick up
the default media padding. For widgets inside the drawer that should
avoid safe areas, developers can add a SafeArea widget.

Hixie committed:
Provide a navigatorKey property on MaterialApp and WidgetsApp (#13591)

This lets people poke at navigators without having to get their
BuildContext from a build function or State first.

gspencergoog committed:
Add a Dart script to prepare the flutter repo for packaging. (#13548)

This is the first step in a two-step process of moving the package preparation step from a recipe in chrome_infra to a dart script in the flutter repo. This will make it easier to make changes to the process. The second step is to change the infra recipe to call this script.

In addition, I added a step to the packaging process to run flutter create for each type of template so that any pub dependencies of the templates get added to the cache that gets packaged (and thus users can run flutter create --offline and have it work).

Note that the actual packaging into a "tar" or "zip" file now happens here, so a developer could actually run this script on their machine to create a package.

cbracken committed:
Apply media padding in UserAccountsDrawerHeader (#13619)

Applies top, left, right media padding to avoid placing content within
the safe area insets (e.g. iPhone X notch area in landscape
orientation).

xster committed:
Split one CustomPainter doc paragraph to bullet points (#13590)

* split doc paragraph

* specify shouldRepaint argument nullability

* review

gspencergoog committed:
Switch to using XZ instead of BZip2 for tar archives. (#13620)

XZ is about 50MB smaller, and it works on both Mac and Linux.

cbracken committed:
Apply media padding to snackbar contents (#13623)

Applies any additional bottom, left, and right media padding inside the
snackbar, if present. This accounts for the iPhone X home indicator
widget and horizontal padding for the sensor housing notch in landscape
orientation.

Hixie committed:
Make animating opacity cheaper. (#13583)

No need to update the compositing bits all the time.

cbracken committed:
Require Xcode 9.0 or later for iOS builds (#13608) (#13622)

We now require Xcode 9.0, which ships with the iOS 11 SDK, for iOS
builds. This change does not affect the minimum supported iOS deployment
target, which remains iOS 8.

The iOS 11 SDK adds support for safe area insets, which model the status
bar, iPhone X home indicator, and symmetric horizontal insets to avoid
the iPhone X sensor housing when in landscape mode.

This re-lands flutter/flutter#13608 now that the devicelab Macs have
been updated to Xcode 9.

xster committed:
update xcode project (#13587)

cbracken committed:
Apply media padding to Contacts demo (#13628)

Applies horizontal safe area insets to the Contacts demo in the Gallery.
This is to support the iPhone X sensor housing notch and other similarly
creative display features when in landscape orientation.

szakarias committed:
Remove app.flx from apk (#13647)

cbracken committed:
Use idevice_id for devicelab iOS device listing (#13624)

This patch migrates iOS device listing from using Xcode instruments to
using the libimobiledevice tools idevice_id and ideviceinfo.

ideviceinfo was previously incompatible with iOS 11 physical devices;
this has now been fixed.

In 58fe8237d2ec012ea51a673bc228f6545af8bb3c flutter_tools migrated from
libimobiledevice-based device listing on iOS to using Xcode instruments
to work around the lack of support for iOS 11. Using instruments entails
several downsides, including a significantly higher performance hit, and
leaking hung DTServiceHub processes in certain cases when a simulator is
running, necessitating workarounds in which we watched for, and cleaned
up leaked DTServiceHub processes. This patch returns reverts the move to
instruments now that it's no longer necessary.

This reverts commit 58fe8237d2ec012ea51a673bc228f6545af8bb3c.

szakarias committed:
Remove app.flx on iOS (#13648)

szakarias committed:
Remove app.flx and add flutter_assets to project.pbxproj of all example apps (#13649)

cbracken committed:
Use idevice_id for devicelab iOS device lookups (#13652)

This replaces the use of ideviceinfo (whose output was then filtered for
device IDs) with an invocation idevice_id -l, which returns the list
directly, one per line.

mravn-google committed:
Switch from app.flx to flutter_assets (#13656)

mravn-google committed:
Bump test timeout threshold (#13661)

gspencergoog committed:
Add support for 7Zip to the packaging script. (#13659)

A modern version of 7Zip (7za.exe) is now installed on the bots, this makes use of that.

kangwang1988 committed:
Merge branch 'master' of github.com:kangwang1988/flutter

* 'master' of github.com:kangwang1988/flutter: (57 commits)
  Refactor for style & documentation.
  [Modified]:Fix analyze issues.
  [Modified]:Add test for skip pod install if possible.
  [Modified]:Refactor codes & documentation according to the flutter team suggestions.
  Fix for incorrect transparent -> DividerColor animation for ExpansionTile (#13449)
  Fix template for the driver test file created by 'flutter create' (#13472)
  Roll engine to edb0201fa258d6a2c05899c185a4e5cba0dc26a3 (#13473)
  Remove bottom padding in SliverAppBar (#13471)
  Have the framework in charge of scheduling frames. (#13344)
  Make it possible to specify multiple test arguments in test.dart (#13383)
  docs to redirect a user to dart:io.HttpClientOverrides (#13465)
  Make create command use pub in offline mode. (#13462)
  Fix tapping on a test in flutter run (#13397)
  Make time picker accessible (#13152)
  [flutter_tools][plugins] generated registerWith guards against re-registration (#13384)
  Let translucent Cupertino bars have its scaffold children automatically pad their heights - second try (#13440)
  Upgrade dartdoc to 0.15.0+1 (#13432)
  Roll engine to 8e39760168e7669a023d1746823d9ab6db4e5df8 (#13442)
  Roll engine to 1cdaab88e33f564eee8455c75479706293f2b655 (#13441)
  a11y cursor movement (#13405)
  ...

kangwang1988 commented:
@xster  
Sorry to bother you. As the company requests to remove sensitive informations including my email before signing a CLA, I rebase the commit history. Please let me know if there is something wrong.

cbracken committed:
Apply media padding in Gallery Pesto demo (#13668)

Applies horizontal safe area insets to the Pesto demo in the Gallery.
This is to support the iPhone X sensor housing notch and other similarly
creative display features when in landscape orientation.

xster commented:
No worries. 
Seems like the new file came from https://github.com/flutter/flutter/pull/13541. 

Looks like you `git merge`ed up to the main branch from 10 days ago instead of `git rebase`ing from head. Might be cleaner if you reset your local branch to c93fef3, cherry picked in your new 0d32875 commit and then rebase on `upstream/master` (don't forget to fetch upstream first).

At that point, you can fix the packages/flutter_tools/lib/src/commands/inject_plugins.dart file.

xster committed:
Add flutter_assets and *.class to gitignore (#13664)

Commit:
Added five new locales to flutter_localizations (#13621)

goderbauer committed:
IconButton is a button, semantically (#13674)

* IconButton is a button, semantically

* fix datepicker test

yjbanov committed:
use different delimiter for generated code; print generated code into mismatch description (#13667)

Hixie committed:
Fix niggling PaginatedDataTable bugs (#13556)

Prevent header from thinking it can wrap and then overflowing.

Fix default footer string which lost its colon (localized values are fine).

Make the "rows per page" drop-down include at least one value even when the table lacks many items. (Previously it would assert if your table was too short.)

Make the footer scrollable.

Fix some todos and improve some debug output.

Tests for much of the above.

alanrussian committed:
Change async stubbing to use thenAnswer. (#13521)

* Change async stubbing to use thenAnswer.

Mockito now prohibits calling thenReturn with Futures and Streams. dart-lang/mockito#79

* Update all Mockito deps to 3.0.0.

* Revert "Update all Mockito deps to 3.0.0."

This reverts commit e8ab9d37c33d3d7fe384abde64ea5b4d72623c75.

I did not correctly update the mockito dep, and there's no easy way to update to 3.0 alpha right now.

* Change thenAnswer((_) => to thenAnswer((invocation) =>

* Add Invocation type to thenAnswer lambdas

Commit:
Gen localization classes intead of a big map (#13653)

* Gen localization classes intead of a big map

* tighten up the newlines so that std output matches file output

* restore locale sorting

goderbauer committed:
selected ListTile is selected (#13676)

* selected ListTile is selected

* review comment

* analyzer fix

goderbauer committed:
Update project.pbxproj of flutter_gallery (#13678)

kangwang1988 committed:
Fix analyze problem for inject_plugins.dart.

kangwang1988 committed:
Merge branch 'master' into master

kangwang1988 committed:
Merge branch 'master' into master

xster commented:
Please remember to reply saying `I signed it!` here when you go through the CLA signing (with any email). For our bots. 

Let me know when you're ready and I'll merge it for you.

kangwang1988 commented:
Thanks a lot. I'm on the way for a CLA. After that, I will let you know.

2017-12-20 14:27 GMT+08:00 xster <notifications@github.com>:

> Please remember to reply saying I signed it! here when you go through the
> CLA signing (with any email). For our bots.
>
> Let me know when you're ready and I'll merge it for you.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/13374#issuecomment-352976685>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAx6u-OGWCjyIAlPkM3hjt16yq-ZJ_nXks5tCKjigaJpZM4Q3Rni>
> .
>

kangwang1988 commented:
I signed it!

xster commented:
Not sure if the cla bot is supposed to re-run. The contributor's employer is in the approved corporate list and the pull request sender and git commiter is in his employer's Authorized Contributors Group. I assume it's ok to merge. 

Thanks for your contribution @kangwang1988

xster commented:
Sorry, have to revert. We'll need to keep the env variable in for the time being for previously created projects. 

Side note: we should probably do #12573 soon.
